src/pluggy/hooks.py

```diff
@@ -163,22 +151,41 @@ def varnames(func):
         except Exception:
             return ()
 
-    try:  # func MUST be a function or method here or we won't parse any args
-        spec = _getargspec(func)
-    except TypeError:
-        return (), ()
+    # func MUST be a function or method here or we won't parse any args
+    # this is what we return if it is not
+    non_function = (), ()
 
-    args, defaults = tuple(spec.args), spec.defaults
-    if defaults:
-        index = -len(defaults)
-        args, defaults = args[:index], tuple(args[index:])
+    if hasattr(inspect, "signature"):
```

Comment:
Can we keep this logic in their own methods like it was before with `_getargspec`? This isolates the hacks around `inspect.signature` which improves readability.